src/GSoft.Extensions.Http.Authentication.ClientCredentialsGrant.Tests/ClientCredentialsTokenHttpMessageHandlerTests.cs

```diff
@@ -0,0 +1,160 @@
+        };
+
+        var ex = await Assert.ThrowsAsync<HttpRequestException>(() => this._clientCredentialsHttpClient.GetStringAsync("https://whatever", CancellationToken.None));
+        Assert.Equal(HttpStatusCode.Unauthorized, ex.StatusCode);
+
+        A.CallTo(() => this._tokenManagementService.GetAccessTokenAsync(TestClientName, CachingBehavior.PreferCache, A<CancellationToken>._)).MustHaveHappenedOnceExactly();
+        A.CallTo(() => this._tokenManagementService.GetAccessTokenAsync(TestClientName, CachingBehavior.ForceRefresh, A<CancellationToken>._)).MustHaveHappenedOnceExactly();
+    }
+
+    [Theory]
+    [InlineData(HttpStatusCode.InternalServerError)]
+    [InlineData(HttpStatusCode.Forbidden)]
+    [InlineData(HttpStatusCode.BadRequest)]
+    [InlineData(HttpStatusCode.BadGateway)]
+    [InlineData(HttpStatusCode.RequestTimeout)]
+    [InlineData(HttpStatusCode.ServiceUnavailable)]
+    public async Task SendAsync_When_Unexpected_Response_Other_Than_Unauthorized_Returns_Unexpected_Response_And_Skips_Second_Retry(HttpStatusCode nonRetriableStatusCode)
+    {
+        this._mockPrimaryHttpMessageHandler.ExpectedHttpResponseMessages = new[]
+        {
+            CreateHttpResponseMessage(nonRetriableStatusCode, "Something wrong happened"),
+        };
+
+        var ex = await Assert.ThrowsAsync<HttpRequestException>(() => this._clientCredentialsHttpClient.GetStringAsync("https://whatever", CancellationToken.None));
+        Assert.Equal(nonRetriableStatusCode, ex.StatusCode);
+
+        A.CallTo(() => this._tokenManagementService.GetAccessTokenAsync(TestClientName, CachingBehavior.PreferCache, A<CancellationToken>._)).MustHaveHappenedOnceExactly();
+        A.CallTo(() => this._tokenManagementService.GetAccessTokenAsync(TestClientName, CachingBehavior.ForceRefresh, A<CancellationToken>._)).MustNotHaveHappened();
+    }
+
+    private static HttpResponseMessage CreateHttpResponseMessage(HttpStatusCode statusCode, string content)
+    {
+        return new HttpResponseMessage(statusCode) { Content = new StringContent(content) };
+    }
+
+    public void Dispose()
+    {
+        this._mockPrimaryHttpMessageHandler.Dispose();
+        this._clientCredentialsHttpClient.Dispose();
+    }
+
+    private sealed class MockPrimaryHttpMessageHandler : DelegatingHandler
+    {
+        private int _currentResponseIndex;
+
+        public MockPrimaryHttpMessageHandler()
+        {
+            this.ExpectedHttpResponseMessages = Array.Empty<HttpResponseMessage>();
+            this._currentResponseIndex = 0;
+        }
+
+        public HttpResponseMessage[] ExpectedHttpResponseMessages { private get; set; }
+
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Assert.Equal(TestAccessToken, request.Headers.Authorization?.Parameter);
+            return Task.FromResult(this.ExpectedHttpResponseMessages[this._currentResponseIndex++]);
```

Comment:
Yeah I did the first version with a "reset"-kind of feature but I removed it because every test creates its own new instance of that "mock http message handler" anyway, internal state is never preserved